core/rapid_test.go

```diff
@@ -0,0 +1,406 @@
+				view *proto.View,
+			) *proto.Message {
+				return buildBasicPreprepareMessage(
+					proposal,
+					proposalHash,
+					certificate,
+					nodes[nodeIndex],
+					view,
+				)
+			}
+
+			// Make sure the prepare message is built correctly
+			backend.buildPrepareMessageFn = func(proposal []byte, view *proto.View) *proto.Message {
+				return buildBasicPrepareMessage(proposalHash, nodes[nodeIndex], view)
+			}
+
+			// Make sure the commit message is built correctly
+			backend.buildCommitMessageFn = func(proposal []byte, view *proto.View) *proto.Message {
+				return buildBasicCommitMessage(proposalHash, committedSeal, nodes[nodeIndex], view)
+			}
+
+			// Make sure the round change message is built correctly
+			backend.buildRoundChangeMessageFn = func(
+				proposal []byte,
+				certificate *proto.PreparedCertificate,
+				view *proto.View,
+			) *proto.Message {
+				return buildBasicRoundChangeMessage(proposal, certificate, view, nodes[nodeIndex])
+			}
+
+			// Make sure the inserted proposal is noted
+			backend.insertBlockFn = func(proposal []byte, _ []*messages.CommittedSeal) {
+				insertedProposals.insertProposal(nodeIndex, proposal)
+			}
+
+			// Make sure the proposal can be built
+			backend.buildProposalFn = func(u uint64) []byte {
+				return proposal
+			}
+		}
+
+		// Initialize the backend and transport callbacks for
+		// each node in the arbitrary cluster
+		backendCallbackMap := make(map[int]backendConfigCallback)
+		transportCallbackMap := make(map[int]transportConfigCallback)
+
+		for i := 0; i < int(numNodes); i++ {
+			i := i
+			backendCallbackMap[i] = func(backend *mockBackend) {
+				commonBackendCallback(backend, i)
+			}
+
+			transportCallbackMap[i] = commonTransportCallback
+		}
+
+		// Create the mock cluster
+		cluster := newMockCluster(
+			numNodes,
+			backendCallbackMap,
+			nil,
+			transportCallbackMap,
+		)
+
+		// Set a small timeout, because of situations
+		// where the byzantine node is the proposer
+		cluster.setBaseTimeout(time.Second * 2)
+
+		// Set the multicast callback to relay the message
+		// to the entire cluster
+		multicastFn = func(message *proto.Message) {
+			cluster.pushMessage(message)
+		}
+
+		// Run the sequence up until a certain height
+		for height := uint64(0); height < desiredHeight; height++ {
+			// Start the main run loops
+			cluster.runSequence(height)
+
+			// Wait until Quorum nodes finish their run loop
+			ctx, cancelFn := context.WithTimeout(context.Background(), time.Second*5)
+			if err := cluster.awaitNCompletions(ctx, int64(quorumOptimal(numNodes))); err != nil {
+				t.Fatalf(
+					fmt.Sprintf(
+						"unable to wait for nodes to complete, %v",
+						err,
+					),
+				)
+			}
+
+			// Shutdown the remaining nodes that might be hanging
+			cluster.forceShutdown()
+			cancelFn()
+		}
+
+		// Make sure that the inserted proposal is valid for each height
+		for _, proposalMap := range insertedProposals.proposals {
```

Comment:
Just to be on the safe side, could you check that the test works and reproducible? eg. by hands introduce a bug into the consensus get a red test and re-run rapid with given seed?

Reply:
What do you mean?

For each block height, we wait for the `quorum` set of nodes to reach consensus (the majority), otherwise the test fails:
https://github.com/0xPolygon/go-ibft/blob/b305b6dcadf2904b20fdfa3d5e7862003cb28014/core/rapid_test.go#L385

I would even suggest that this proposal checking is not required at this point, since the test would never complete gracefully if the proposal was invalid for the quorum nodes

Reply:
I mean that now you have a green test, although we don't know:
1. is it useful
2. is it reproducible

I suggest to manually introduce a bug to the code and check that the test become "red", and then that with given seed it's possible to reproduce the failing test.

Reply:
The test will fail if over 1/3 of nodes are byzantine:
https://github.com/0xPolygon/go-ibft/blob/b305b6dcadf2904b20fdfa3d5e7862003cb28014/core/rapid_test.go#L233

You can introduce this manually in the test by hardcoding `numByzantineNodes` to `maxFaulty(numNodes) + 1`.

Keep in mind both tests introduced in this PR are migrated from `pbft-consensus`, where the definition of a passing test [was having 2/3 + 1 nodes reach consensus](https://github.com/0xPolygon/pbft-consensus/blob/547d3058f51ad510ea5495f496f8ed66d74cb1b4/e2e/rapid_test.go#L104-L107).

We can talk about adding tests that specifically cause a cluster to not reach consensus, but they are out of scope for this task, which was only migration